packages/portfolio-contract/src/portfolio.exo.ts

```diff
@@ -39,195 +45,187 @@ export const DECODE_CONTRACT_CALL_RESULT_ABI = [
+            if (message.success) {
+              const [address] = decodeAbiParameters(
+                [{ type: 'address' }],
+                message.result,
               );
+              const { Gmp } = this.state;
+              if (Gmp) {
+                // TODO: remoteAddressVK is set only once
+                Gmp.remoteAddressVK.resolver.resolve(address);
+              }
+              trace(`remoteAccountAddress ${address}`);
             }
           }
-
+          // TODO: Handle the result of the contract call
           trace('receiveUpcall completed');
         },
       },
       keeper: {
         getLCA() {
-          const { localAccount } = this.state;
-          return localAccount;
+          const { Gmp } = this.state;
+          if (!Gmp) throw Error('Gmp not set');
+          return Gmp.localAccount;
         },
         getPositions(): YieldProtocol[] {
           const { state } = this;
           return harden(
             (keys(YieldProtocol) as YieldProtocol[]).filter(k => !!state[k]),
           );
         },
+        getUSDNICA() {
+          const { USDN } = this.state;
+          if (!USDN) throw Error('USDN not set');
+          return USDN;
+        },
+        getRemoteAddress() {
+          const { Gmp } = this.state;
+          if (!Gmp) throw Error('Gmp not set');
+          return Gmp.remoteAddressVK.vow;
+        },
         initAave(chain: CaipChainId) {
           this.state.Aave = harden({
             chain,
-            remoteAddressVK: vowTools.makeVowKit(),
           });
-          trace('initAave', chain);
         },
-        initCompound() {
+        initCompound(chain: CaipChainId) {
           this.state.Compound = harden({
-            chain: EthL1,
-            remoteAddressVK: vowTools.makeVowKit(),
+            chain,
           });
-          trace('initCompound');
         },
         initUSDN(account: NobleAccount) {
           this.state.USDN = account;
-          trace('initUSDN', account);
         },
-        getUSDNICA() {
-          const { USDN } = this.state;
-          if (!USDN) throw Error('USDN not set');
-          return USDN;
-        },
-        getAaveAddress() {
-          const { Aave } = this.state;
-          if (!Aave) throw Error('Aave not set');
-          return Aave.remoteAddressVK.vow;
+
+        initGmp(account: LocalAccount) {
+          this.state.Gmp = harden({
+            localAccount: account,
+            remoteAddressVK: vowTools.makeVowKit(),
+          });
+          trace('initLCA');
         },
-        getCompoundAddress() {
-          const { Compound } = this.state;
-          if (!Compound) throw Error('Compound not set');
-          return Compound.remoteAddressVK.vow;
+        wait(val: bigint) {
+          return vowTools.watch(E(timer).delay(val));
         },
       },
       rebalanceHandler: {
-        async handle(seat: ZCFSeat, offerArgs: CopyRecord) {
+        async handle(seat: ZCFSeat, offerArgs: EVMOfferArgs) {
```

Comment:
hm... as the [Offer Arguments docs](https://docs.agoric.com/guides/zoe/proposal.html#offer-arguments) say, we can't rely on the caller to pass us correct data. So I might suggest...

```suggestion
        async handle(seat: ZCFSeat, offerArgs: unknown) {
```

But we do have a typed `mustMatch` before we use it. And it's convenient when _we_ are the caller to get static typing info.

We just have to be very dilligent to not add any code that uses `offerArgs` between the method declaration and the `mustMatch`.